### ISSUE TITLE
Optimise feeding of batches

### ISSUE THREAD

**Issue by RustyPanda** [RustyPanda/zoobot#20]
Reading in examples feels not-quite-right.

Currently, I convert the batched dataset to an iterator, call it for a batch, then feed this batch to stratified_sample. The output of stratified_sample becomes the model input.

In practice, this means linking together various queues. 

1. I think it might be possible to call the sampler directly with the batch iterator: I should see what happens

2. I'm not clear on the difference between rejection and stratified sampling.

[The guide](https://www.tensorflow.org/api_guides/python/contrib.training#Online_data_resampling) says:

> "To resample data with replacement on a per-example basis, use tf.contrib.training.rejection_sample. [...] Use tf.contrib.training.stratified_sample to resample without replacement."

Resampling without replacement causes me to burn through epochs much faster than I'm actually *using* data. If pre-processing is being applied, this is inefficient - but I think it's probably not. *If I stick with stratified, I should set the dataset to repeat n times.*

Resampling with replacement would be more efficient if pre-processing is applied, but
- I don't think pre-processing is applied
- I don't know what happens when a dataset finishes

*I should first test if pre-processing is being applied*

https://www.tensorflow.org/api_docs/python/tf/contrib/training/rejection_sample
Accept function can be init_probs[label]

Faster rejection sampler: https://github.com/tensorflow/tensorflow/pull/18730/commits/b1067116c6a2351f4c597a9391b21ad0f513565b

**Comment by RustyPanda** [comment]
On GPU, I might need to rewrite whatever I decide on to use a pure tf graph. Will open separate issue.
